agent/processwatcher.go

```diff
@@ -0,0 +1,255 @@
+package agent
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/appcanary/libspector"
+)
+
+type ProcessWatcher interface {
+	Start()
+	Stop()
+	Match() string
+	State() *watchedState
+}
+
+type processWatcher struct {
+	sync.Mutex
+	keepPolling  bool
+	UpdatedAt    time.Time
+	OnChange     ChangeHandler
+	pollSleep    time.Duration
+	BeingWatched bool
+	match        string
+	state        *watchedState
+}
+
+type watchedState map[int]watchedProcess
+
+func (w watchedState) String() string {
+	buffer := bytes.NewBufferString("Watched Processes:\n\n")
+	for _, proc := range w {
+		buffer.WriteString(fmt.Sprintf("%v\n", proc.String()))
+	}
+	return buffer.String()
+}
+
+type watchedProcess struct {
+	ProcessStarted time.Time
+	Libraries      []library
+	Outdated       bool
+	Pid            int
+}
+
+func (w *watchedProcess) String() string {
+	buffer := bytes.NewBufferString(fmt.Sprintf("PID: %d", w.Pid))
+	if w.Outdated {
+		buffer.WriteString(", is running outdated lib(s)")
+	}
+	for _, lib := range w.Libraries {
+		buffer.WriteString(fmt.Sprintf("\n%v", lib.String()))
+	}
+	return buffer.String()
+}
+
+type library struct {
+	SpectorLib libspector.Library
+	Outdated   bool
+}
+
+func (l *library) String() string {
+	buffer := bytes.NewBufferString("")
+	if l.Outdated {
+		buffer.WriteString("Outdated: yes")
+	} else {
+		buffer.WriteString("Outdated:  no")
+	}
+	buffer.WriteString(fmt.Sprintf(", Path: %v, ", l.SpectorLib.Path()))
+	if pkg, err := l.SpectorLib.Package(); err != nil {
+		buffer.WriteString(fmt.Sprintf("Package error: %v", err))
+	} else {
+		buffer.WriteString(fmt.Sprintf("Package: %v, ", pkg))
+	}
+	if modified, err := l.SpectorLib.Modified(); err != nil {
+		buffer.WriteString(fmt.Sprintf("Modified error: %v", err))
+	} else {
+		buffer.WriteString(fmt.Sprintf("Modified: %v", modified))
+	}
+	return buffer.String()
+}
+
+func (wp *watchedProcess) MarshalJSON() ([]byte, error) {
+	libs := make([]map[string]interface{}, len(wp.Libraries))
+
+	for i, lib := range wp.Libraries {
+		path := lib.SpectorLib.Path()
+
+		modified, err := lib.SpectorLib.Modified()
+		if err != nil {
+			log.Warningf("error retrieving modification date for lib %s, %v", path, err)
+			continue
+		}
+
+		var fullPackage string
+		if pkg, err := lib.SpectorLib.Package(); err != nil {
+			log.Warningf("error retrieving package name for lib %s, %v", path, err)
+		} else {
+			fullPackage = fmt.Sprintf("%s-%s", pkg.Name(), pkg.Version())
```

Comment:
this may or may or not matter but make sit a bit simpler on receiving end if we send "package_name" and "package_version"